### PR TITLE
Use syncronized to make the loadbalancer semantically thread safe.

### PR DESCRIPTION
**Pull request by baldersheim**
Use a cache to avoid parsing a string to get an index.
Move test to same package to avoid public access to internal details.

@vekterli PR
@yngveaasheim FYI

### CODE REVIEW

File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancer.java [hunk 1]
```diff
@@ -68,24 +77,24 @@ public Node getRecipient(List<Mirror.Entry> choices) {
...
+                NodeMetrics nodeMetrics = getNodeMetrics(entry);
 
-            weightSum += nodeMetrics.weight;
+                weightSum += nodeMetrics.weight;
 
-            if (weightSum > position) {
-                selectedNode = new Node(entry, nodeMetrics);
-                break;
+                if (weightSum > position) {
+                    selectedNode = new Node(entry, nodeMetrics);
```
Comment (vekterli) [inline documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancer.java:87]: The returned `NodeMetrics` object is shared and thus may be read outside the lock whilst concurrently being updated by another thread. Is this a problem? Should we deep-copy the metrics to make sure they're internally consistent? This is not something introduced by this PR, but just a general observation.

**Review by vekterli — APPROVED** [review]
:+1: I need some time to digest the existing algorithm for this, but synchronization changes look good and also should make it easier to reason about the behavior in general.

File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancer.java (hunk 1 above)
Comment (baldersheim) [reply documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancer.java:87]: That was why I made it package private. I checked all other usage. That was only in tests.

File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancer.java [hunk 2]
```diff
@@ -54,6 +55,14 @@ public int getIndex(String nodeName) {
             throw new IllegalArgumentException(err, e);
         }
     }
+    private int getCachedIndex(String nodeName) {
+        Integer index = cachedIndex.get(nodeName);
+        if (index == null) {
+            index = getIndex(nodeName);
+            cachedIndex.put(nodeName, index);
+        }
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancer.java]: Done, I think the JIT compiler will solve the performance for us.

**Pull request merged**